genvm/rewards.go

```diff
@@ -2,79 +2,107 @@ package vm
-	byCoinbase := make(map[types.Address]util.Weight)
+	unique := map[core.Address]struct{}{}
 	for _, reward := range rewards {
-		weight := util.WeightFromNumDenom(reward.Weight.Num, reward.Weight.Denom)
-		logger.With().Debug("coinbase weight", reward.Coinbase, log.Stringer("weight", weight))
-		totalWeight.Add(weight)
-		if _, ok := byCoinbase[reward.Coinbase]; ok {
-			byCoinbase[reward.Coinbase].Add(weight)
-		} else {
-			byCoinbase[reward.Coinbase] = weight
+		if reward.Weight.Num == 0 || reward.Weight.Denom == 0 {
+			return fmt.Errorf("reward with invalid (zeroed) weight (%d/%d) included into the block for %v", reward.Weight.Num, reward.Weight.Denom, reward.Coinbase)
 		}
+		if _, exists := unique[reward.Coinbase]; exists {
+			return fmt.Errorf("multiple rewards for the same coinbase %v", reward.Coinbase)
+		}
+		unique[reward.Coinbase] = struct{}{}
 	}
-	if totalWeight.Cmp(util.WeightFromUint64(0)) == 0 {
-		logger.Error("zero total weight in block rewards")
-		return nil, fmt.Errorf("zero total weight")
+	return nil
+}
+
+func (v *VM) addRewards(lctx ApplyContext, ss *core.StagedCache, tx *sql.Tx, fees uint64, blockRewards []types.AnyReward) error {
+	var (
+		layersAfterEffectiveGenesis = lctx.Layer.Difference(types.GetEffectiveGenesis())
+		subsidy                     = erewards.TotalSubsidyAtLayer(layersAfterEffectiveGenesis)
+		total                       = subsidy + fees
+		transferred                 uint64
+		totalWeight                 = util.WeightFromUint64(0)
+	)
+	for _, blockReward := range blockRewards {
+		totalWeight.Add(util.WeightFromNumDenom(blockReward.Weight.Num, blockReward.Weight.Denom))
 	}
-	finalRewards := make([]*types.Reward, 0, len(rewards))
-	layerRewards := calculateLayerReward(cfg)
-	totalRewards := layerRewards + totalFees
-	logger.With().Debug("rewards info for layer",
-		log.Uint64("layer_rewards", layerRewards),
-		log.Uint64("fee", totalFees))
-	rewardPer := util.WeightFromUint64(totalRewards).Div(totalWeight)
-	lyrRewardPer := util.WeightFromUint64(layerRewards).Div(totalWeight)
-	seen := make(map[types.Address]struct{})
-	for _, reward := range rewards {
-		if _, ok := seen[reward.Coinbase]; ok {
-			continue
+	for _, blockReward := range blockRewards {
+		relative := util.WeightFromNumDenom(blockReward.Weight.Num, blockReward.Weight.Denom).Div(totalWeight)
+
+		totalReward := new(big.Int).SetUint64(total)
+		totalReward.
+			Mul(totalReward, relative.Num()).
+			Quo(totalReward, relative.Denom())
+		if !totalReward.IsUint64() {
+			return fmt.Errorf("%w: total reward %v for %v overflows uint64",
+				core.ErrInternal, totalReward, blockReward.Coinbase)
+		}
+
+		subsidyReward := new(big.Int).SetUint64(subsidy)
+		subsidyReward.
+			Mul(subsidyReward, relative.Num()).
+			Quo(subsidyReward, relative.Denom())
+		if !subsidyReward.IsUint64() {
+			return fmt.Errorf("%w: subsidy reward %v for %v overflows uint64",
+				core.ErrInternal, subsidyReward, blockReward.Coinbase)
```

Comment:
looks like you can factor out an utility method to do the division and check uint64

Reply:
it is much better without such method